### DISCUSSION TITLE
Seems window.__TAURI_IPC__ only work in site of devPath

### DISCUSSION THREAD

**Discussion by linonetwo** [tauri-apps/tauri#6033] [Q&A] [answered]
In `tauri.conf.json -> build.devPath` we have `"devPath": "http://localhost:1420",` by default. And tauri ipc methods injected as expected in this website.

But if we ask tauri to load another dev site, using redirect like:

```jsx
export function DevPage() {
  useEffect(() => {
    location.href = 'http://localhost:8080';
    redirect('http://localhost:8080');
  }, []);
  return null;
}

export function BasicPage() {
  useEffect(() => {
    location.href = '/out/index.html';
    redirect('/out/index.html');
  }, []);
  return null;
}

export const routes = [
  {
    path: '/',
    name: 'Landing Page',
    element: <LandingPage />,
  },
  {
    path: '/basic',
    name: 'Basic',
    element: <BasicPage />,
  },
  {
    path: '/dev',
    name: 'Dev',
    element: <DevPage />,
  },
];
export const router = createBrowserRouter([
  { path: '/', element: <RootLayout />, children: routes },
]);
```

I found it will error about `TypeError: window.__TAURI_IPC__ is not a function`.

And if I change ``tauri.conf.json -> build.devPath` to `"devPath": "http://localhost:8080",`, then this error gones. But I will lost my landing page on dev mode (I use it to choose which dev site I'm connection to).

And the `/basic`route in landing page will redirect to a static HTML file in the /public folder, in which, the `window.__TAURI_IPC__` injection works.

So I can guess, there are some hidden security check, that disable the `window.__TAURI_IPC__` injection, when a localhost site is not listed in the `devPath`, but if the webpage is a static HTML file it still works, right?

---

My main.rs is:

```rust
#![cfg_attr(
  all(not(debug_assertions), target_os = "windows"),
  windows_subsystem = "windows"
)]

mod commands;
mod state;
use dotenvy::dotenv;
use state::AppState;
use std::env;
use tauri::TitleBarStyle;
use tokio::sync::Mutex;

#[tokio::main]
async fn main() {
  tauri::async_runtime::set(tokio::runtime::Handle::current());
  dotenv().ok();
  let preload = include_str!("../../public/preload/index.js");
  let is_dev = env::var("NODE_ENV").unwrap_or_default() == "development";
  let initial_path = if is_dev {
    "index.html"
  } else {
    "out/index.html"
  };
  tauri::Builder::default()
    .manage(AppState(Mutex::new(
      state::AppStateRaw::new().await.unwrap(),
    )))
    // manually create window here, instead of in the tauri.conf.json, to add `initialization_script` here
    .setup(move |app| {
      let _window =
        tauri::WindowBuilder::new(app, "label", tauri::WindowUrl::App(initial_path.into()))
          .title("XXX")
          .inner_size(1000.0, 800.0)
          .title_bar_style(TitleBarStyle::Overlay)
          .hidden_title(true)
          .initialization_script(&preload)
          .build()?;
      Ok(())
    })
    .invoke_handler(commands::invoke_handler())
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
```

**Comment by FabianLars** [comment, marked as answer]
> So I can guess, there are some hidden security check, that disable the window.__TAURI_IPC__ injection, when a localhost site is not listed in the devPath, but if the webpage is a static HTML file it still works, right?

Yep, pretty much that. Tauri's apis are only injected into devPath and distDir, and it "breaks" on redirects. This is by design but an option to inject in more pages is being worked on here https://github.com/tauri-apps/tauri/pull/5918
